File: python/209.py
# Problem 209: Minimum Size Subarray Sum
# Difficulty: Medium
# Category: Array, Two Pointers, Binary Search
# Link: https://leetcode.cn/problems/minimum-size-subarray-sum/
#
# 题目描述：
# 给定一个含有 n 个正整数的数组和一个正整数 target 。
# 找出该数组中满足其和 ≥ target 的长度最小的 连续子数组 [numsl, numsl+1, ..., numsr-1, numsr] ，并返回其长度。如果不存在符合条件的子数组，返回 0 。

from typing import List

class Solution:
    def minSubArrayLen(self, target: int, nums: List[int]) -> int:
        n = len(nums)                # ❶ 获取数组总长度（菜市场总摊位数量）
        head, sums, min_sub = 0, 0, float('inf') # ❷ 初始化：蛇头位置/吃进的食物量/最小记录本

        for tail in range(n):        # ❸ 蛇尾开始逐个摊位尝鲜（右指针移动）
            sums += nums[tail]       # ❹ 吞下当前摊位的食物（累加数字）
            
            while sums >= target:    # 🌟 关键点：当吃饱时开始收缩身体
                min_sub = min(min_sub, tail - head + 1) # ❺ 记录当前身体长度（尾巴-头+1）
                sums -= nums[head]   # ❻ 吐出最前面吃的那份食物（减掉头部数字）
                head += 1            # ❼ 蛇头向前缩进（左指针右移）

        return min_sub if min_sub != float('inf') else 0 # ❽ 返回记录本上的最小值
    

if __name__ == '__main__':
    testcase = [1, 2, 3, 3, 1, 4, 4, 5]
    target = 9
    print(Solution().minSubArrayLen(target, testcase))

File: ruby/209.rb
# Problem 209: Minimum Size Subarray Sum
# Difficulty: Medium
# Category: Array, Two Pointers, Binary Search
# Link: https://leetcode.cn/problems/minimum-size-subarray-sum/
#
# 题目描述：
# 给定一个含有 n 个正整数的数组和一个正整数 target 。
# 找出该数组中满足其和 ≥ target 的长度最小的 连续子数组 [numsl, numsl+1, ..., numsr-1, numsr] ，并返回其长度。如果不存在符合条件的子数组，返回 0 。
#
# @param {Integer} target
# @param {Integer[]} nums
# @return {Integer}
def min_sub_array_len(target, nums)
  head, sum = 0, 0 # 初始化头指针，连续字数组之和
  res = nums.size + 1 # 初始化满足条件的连续字数组长度，将其设定为nums.size中不可能出现的一个数，且必须大于nums.size

  nums.size.times do |tail| # 开始相加
    sum += nums[tail] # 循环相加连续子数组的和
    while sum >= target # 关键点：当和大于等于目标时候，开始滑动窗口骚操作
      res = [res, tail - head + 1].min # # 比对最小的长度，赋值给结果，为什么tail-head+1? 如果此时tail位于4，head位于0，则中间有5个元素
       # 骚操作2: 当和大于等于目标时候，将最开始加进去的那个head吐出来，让head指针往后移动。
       # 如果此时sums小于target，则tail自动往后移动，直到sums>target，又会进入while block，重复上一步操作，直到找到一个最小窗口，或者最外层循环结束。
      sums -= nums[head]
      head += 1
    end
  end
  res == nums.size + 1 ? 0 : res # 如果整个数组里没有符合条件的子数组，则res不会走到while循环里，自然就等于初始赋值。
end